substrate/frame/support/src/traits/messages.rs

```diff
@@ -174,6 +176,61 @@ pub struct BatchFootprint {
 	pub new_pages_count: u32,
 }
 
+/// The resource footprints of continuous subsets of messages.
+///
+/// For a set of messages `xcms[0..n]`, each `footprints[i]` contains the footprint
+/// of the batch `xcms[0..i]`, so as `i` increases `footprints[i]` contains the footprint
+/// of a bigger batch.
+#[derive(Default, RuntimeDebug)]
+pub struct BatchesFootprints {
+	pub first_page_pos: usize,
+	pub footprints: Vec<BatchFootprint>,
+}
+
+impl BatchesFootprints {
+	/// Appends a batch footprint to the back of the collection.
+	///
+	/// The new footprint represents a batch that includes all the messages contained by the
+	/// previous batches plus the provided `msg`. If `new_page` is true, we will consider that
+	/// the provided `msg` is appended to a new message queue page. Otherwise, we consider
+	/// that it is appended to the current page.
+	pub fn push(&mut self, msg: &[u8], new_page: bool) {
+		let previous_footprint =
+			self.footprints.last().map(|footprint| *footprint).unwrap_or_default();
+
+		let mut new_pages_count = previous_footprint.new_pages_count;
+		if new_page {
+			new_pages_count = new_pages_count.saturating_add(1);
+		}
+		self.footprints.push(BatchFootprint {
+			msgs_count: previous_footprint.msgs_count.saturating_add(1),
+			size_in_bytes: previous_footprint.size_in_bytes.saturating_add(msg.len()),
+			new_pages_count,
+		});
+	}
+
+	/// Gets the biggest batch for witch the comparator function returns `Ordering::Less`.
+	pub fn search_best_by<F>(&self, f: F) -> &BatchFootprint
+	where
+		F: FnMut(&BatchFootprint) -> Ordering,
```

Comment:
I am also thining about something like remove `pub` from `pub first_page_pos` and return it just from `search_best_by`,
but I am not sure if it would make caller's code simpler:

```suggestion
	pub fn search_best_by<F>(&self, f: F) -> (&BatchFootprint, first_page_pos)
	where
		F: FnMut(&BatchFootprint, first_page_pos) -> Ordering,
```

Reply:
Personally I find it easier to understand if we use `BatchesFootprints.first_page_pos` since this variable is for all batches, not just for the `search_best_by`. Another option would be to remove `pub` and define a getter. But I don't know if it's worth it.